### PR TITLE
Grpc object cache

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/GrpcObjectsCacheConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import com.netflix.archaius.api.annotations.DefaultValue;
+
+@Configuration(prefix = "titus.grpc.cache")
+public interface GrpcObjectsCacheConfiguration {
+
+    /**
+     * @return whether or not the cache is enabled
+     */
+    @DefaultValue("true")
+    boolean isGrpcObjectsCacheEnabled();
```
Comment (joshi-keyur) [inline titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/GrpcObjectsCacheConfiguration.java:13]: Should we separate into two properties for jobs and tasks in case we want to disable caching for either one for performance reasons?

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    void shutdown() {
+        scheduleRef.cancel();
+    }
+
+    C get(String id, E entity) {
+        CacheEntry<E, C> entry = cache.get(id);
+        if (entry != null) {
+            if (entry.getCoreEntity() == entity) {
```
Comment (joshi-keyur) [inline titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java:92]: Do you think we should use equality check here? I am curious how many times we would hit the negative of this condition but not a complete cache miss.

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/GrpcObjectsCacheConfiguration.java (hunk 1 above)
Comment (tbak) [reply titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/GrpcObjectsCacheConfiguration.java:13]: I think we should be able to disable cache as a whole in case there is a problem with it. It is very unlikely we might get odd behavior with one entity, and normal with another.

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java (hunk 2 above)
Comment (tbak) [reply titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java:92]: No, it would be too expensive. It is done on purpose like that. JobService will always return the same, latest value. If that value changes (even if equals is true), we should update the cache so we can depend on fast reference comparison.

**Review by andrew-leung — APPROVED** [review]

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/DefaultGrpcObjectsCache.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    public void shutdown() {
+        Evaluators.acceptNotNull(jobCache, ProtobufCache::shutdown);
+        Evaluators.acceptNotNull(taskCache, ProtobufCache::shutdown);
+    }
+
+    public Job getJob(com.netflix.titus.api.jobmanager.model.job.Job<?> coreJob) {
+        if (configuration.isGrpcObjectsCacheEnabled()) {
+            return jobCache.get(coreJob.getId(), coreJob);
+        } else {
+            return toGrpcJob(coreJob);
```
Comment (joshi-keyur) [inline titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/DefaultGrpcObjectsCache.java:90]: If the object cache is disabled, this method invocation would trigger serialization of the `job` object every time `getJob` is called for the `coreJob`. Should this be optimized?

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    private static final String METRICS_ROOT = "protobufCache.";
+
+    private final Function<E, C> converter;
+    private final Supplier<Predicate<E>> expiredPredicateSupplier;
+    private final GrpcObjectsCacheConfiguration configuration;
+    private final Clock clock;
+
+    @VisibleForTesting
+    final ConcurrentMap<String, CacheEntry<E, C>> cache = new ConcurrentHashMap<>();
```
Comment (fabiokung) [inline titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java:49]: did you consider using a Guava cache with capacity-based LRU eviction, instead of relying only on time-based periodic evictions? It would be more resilient to memory blowing up and potentially simpler implementation?

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                cache.remove(entry.getKey());
+            }
+        }
+
+        // Find expired entries
+        Predicate<E> expiredPredicate = expiredPredicateSupplier.get();
+        Set<String> candidates = CollectionsExt.copyAndRemove(cache.keySet(), toBeRemoved.keySet());
+        candidates.forEach(id -> {
+            CacheEntry<E, C> entry = cache.get(id);
+            if (expiredPredicate.test(entry.getCoreEntity())) {
```
Comment (fabiokung) [inline titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java:121]: there's quite a bit of re-work here, since the predicate for each entity will recompute the `all[Job|Task]Ids` `HashMap` for entry. It's probably worth changing the API here to compute `items that should be present` only once to be tested against all entries in the cache.

Alternatively, change to capacity based eviction (instead of periodic time-based cleanup) as I mentioned on a previous comment.

**Review by fabiokung — APPROVED** [review]

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java (hunk 4 above)
Comment (tbak) [reply titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java:49]: I wanted to use Caffeine, but in the end I decided that it will be better to write a custom cache implementation.

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java (hunk 5 above)
Comment (tbak) [reply titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java:121]: We want to minimize calls to JobService API which are relatively expensive. So before checking the expiry of cached entries, we fetch all of them with a single call. Next, we use the cached value (in the clojure), to check every item (this is fast).

Comment (fabiokung) [reply titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java:121]: Yeah, I get the intent, let me rephrase for clarity. The `expiredPredicate` used here (reproduced below) is evaluated for each candidate (all cache entries), and each time it is evaluated, it recomputes the `known[Job|Task]Ids` Set:

```diff
+                () -> {
+                    List<com.netflix.titus.api.jobmanager.model.job.Job> allJobs = jobOperations.getJobs();
+                    Set<String> knownJobIds = new HashSet<>();
+                    allJobs.forEach(job -> knownJobIds.add(job.getId()));
+                    return job -> !knownJobIds.contains(job.getId());
+                },
+
```

I know `jobOperations.getJobs()` is all in-memory and is a lightweight operation, and building a `HashSet` out of it is not expensive. It's probably not a bottleneck, but removing the O(N^2) here and computing `known[Job|Tasks]Ids` set only once every eviction cycle would be a simple enough optimization that is probably worth it.

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/DefaultGrpcObjectsCache.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        if (configuration.isGrpcObjectsCacheEnabled()) {
+            return taskCache.get(coreTask.getId(), coreTask);
+        } else {
+            return toGrpcTask(coreTask);
+        }
+    }
+
+    private Task toGrpcTask(com.netflix.titus.api.jobmanager.model.job.Task coreTask) {
+        Task task = GrpcJobManagementModelConverters.toGrpcTask(coreTask, logStorageInfo);
+        // Need to serialize it once since GeneratedMessageV3 isn't entirely immutable.
```
Comment (fabiokung) [inline titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/DefaultGrpcObjectsCache.java:104]: minor: move the duplicated comment to javadoc of the method being called

File: titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java (hunk 5 above)
Comment (fabiokung) [reply titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java:121]: My bad, I didn't realize we were holding a `Supplier<Predicate>` here, and computing the whole `Predicate` only once before checking each item. Nevermind my comments, LGTM 👍
